File: Leetcode/linkedlist/109_Convert Sorted List to Binary Search Tree/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(head==nullptr) return nullptr;
        if(head->next==nullptr) return new TreeNode(head->val);
        ListNode *slow = head, *fast = head, *prev = head;
        while( fast!=nullptr && fast->next!=nullptr){
            prev = slow;
            fast = fast->next->next;
            slow = slow->next;
        }
        TreeNode* res = new TreeNode(slow->val);
        prev->next = nullptr;
        res->left = sortedListToBST(head);
        res->right = sortedListToBST(slow->next);
        return res;
    }
};

File: Leetcode/linkedlist/109_Convert Sorted List to Binary Search Tree/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if not head: return None
        if not head.next: return TreeNode(head.val)
        slow, fast, prev = head, head, head
        while fast and fast.next:
            prev = slow
            fast = fast.next.next
            slow = slow.next
        res = TreeNode(slow.val)
        prev.next = None
        res.left = self.sortedListToBST(head)
        res.right = self.sortedListToBST(slow.next)
        return res
